### DISCUSSION TITLE
Performance of Reducescatter overlapping with backward pass computation

### DISCUSSION THREAD

**Discussion by maxhgerlach** [horovod/horovod#3819] [General]
I have profiled a training with TensorFlow 2.9.1 that transfers gradients with `hvd.grouped_reducescatter`. When these transfers overlap with backward pass computation ops that are still in progress, I observe some peculiarities that I do not see with a similar model that uses `hvd.grouped_allreduce` instead. The NSight Systems screenshot below contains some Horovod Timeline activities (bottom row), a view of some running CUDA kernels above that, and PCIe/NVLink communication bandwidth in the middle.

![Screenshot 2023-01-09 at 16 04 19](https://user-images.githubusercontent.com/1778667/212874245-e4fc26ac-61e3-4212-a1dd-cad875f0cda3.png)

There is a good chance that improving the following could reduce training step times if more communication could be done before the backward pass is completed.

1. The `MEMCPY_IN_FUSION_BUFFER` step seems to take ~20 msec here. For most of that time there is no kernel running in the Horovod/NCCL stream 24.
2. There is no such gap for the main activity `NCCL_REDUCESCATTER`. The kernel `ncclkernel_ReduceScatter_RING_LL_Sum_half` runs for the entire time. However, it takes ~100 msec to complete, while a typical duration for this size would be ~4 msec when the GPU is not doing much more. Also, for long stretches of this time, no actual transfers appear to be in progress: The bandwidth plots look mostly idle.

(1): I think I may have figured out what's going on with (1). Since TensorFlow output tensor allocations are asynchronous, but happen in a different CUDA stream, the timeline currently doesn't show the actual time of waiting for that buffer to be available here 
 https://github.com/horovod/horovod/blob/ca69f2c521bc5c19f2fb196a3ecfa0bb590b193a/horovod/common/ops/nccl_operations.cc#L1244, rather it's hidden behind MEMCPY_IN_FUSION_BUFFER. This is different with Allreduce, where we know the output buffer size at enqueue time and allocate it then (importantly, this doesn't block the Horovod background thread). It would be great to get rid of these delays. One way would be to add an optional API to provide a hint about the size of the output tensor. If the user knows how large the resulting tensor will be, we could pre-allocate it when the op is enqueued, without any inter-process communication. A similar optimization could also be done for Allgather.

(2): Here I don't know what could be done. Supposedly the CUDA stream Horovod uses for NCCL is created with the highest priority (https://github.com/horovod/horovod/blob/ca69f2c521bc5c19f2fb196a3ecfa0bb590b193a/horovod/common/ops/cuda_operations.cc#L207). Does this priority actually affect all NCCL operations? Or could this also be somehow related to the cross-stream synchronization related to that output buffer allocation? (I haven't seen similar slowdowns with Allreduce...)

**Comment by maxhgerlach** [comment]
After discussing with Nicolas in Slack it seems likely that (2) may be a symptom of not all GPUs being ready to start the NCCL communication at the same time. In particular, if other ranks take even longer to allocate the output buffer, this may add to the delay.

Torch Distributed has `all_gather_into_tensor` and `reduce_scatter_into_tensor`, which write into a preallocated output tensor (https://pytorch.org/docs/stable/distributed.html). A similar API would avoid the problem in Horovod. However, it wouldn't work with TensorFlow where tensors are immutable. 

Introducing an optional `output_shape` argument would be a possibility for Allgather, helping users who know the resulting shape in advance. For Reducescatter, Horovod should actually be able to compute the output tensor shape early without any communication because it's required that all input tensors have the same shape.

**Comment by maxhgerlach** [comment]
Indeed, pulling the Reducescatter output buffer allocations out of the Horovod background thread and having TensorFlow start with them before enqueuing the Horovod op improves everything (#3824). 

The perceived runtime of the NCCL kernel is now well aligned with PCIe and NVLink transfers. The time spent waiting for the input and output tensors to be available is now much better accounted for in Horovod timeline activities as `WAIT_FOR_DATA` and `WAIT_FOR_OTHER_TENSOR_DATA` (and it's shorter). In my quick profiling run, many more Reducescatter ops can be completed before the backward pass ends.

![Screenshot 2023-01-27 at 16 54 03](https://user-images.githubusercontent.com/1778667/215132707-2a88ce9b-87d7-4cea-a246-77ef4797cb2b.png)
